src/modules/job-info/job_state.c

```diff
@@ -0,0 +1,479 @@
+/************************************************************\
+ * Copyright 2018 Lawrence Livermore National Security, LLC
+ * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+ *
+ * This file is part of the Flux resource manager framework.
+ * For details, see https://github.com/flux-framework.
+ *
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+/* job_state.c - store information on state of jobs */
+
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+#include <czmq.h>
+#include <jansson.h>
+#include <flux/core.h>
+
+#include "src/common/libeventlog/eventlog.h"
+
+#include "job_state.h"
+
+/* To handle the common case of user queries on job state, we will
+ * store jobs in three different lists.
+ *
+ * - pending - these are jobs that have not yet in the RUN state, they
+ *   are sorted based on job priority, then job submission time.  -
+ * - running - these are jobs that have transitioned to the RUN state.
+ *   They are sorted by initial run start time.
+ * - inactive - these are jobs that are in the INACTIVE state, they
+ *   are sorted by job completion time.
+ *
+ * There is also an additional list `processing` that stores jobs that
+ * cannot yet be stored on one of the lists above.
+ *
+ * The list `futures` is used to store in process futures.
+ */
+
+struct job_state_ctx {
+    flux_t *h;
+    zhashx_t *index;
+    zlistx_t *pending;
+    zlistx_t *running;
+    zlistx_t *inactive;
+    zlistx_t *processing;
+    zlist_t *futures;
```

Comment:
"One of these things is not like the others; one of these things does not belong..."
Ahem sorry, Sesame Street.  Should the lone zlist_t be a zlistx_t just to avoid having both API's mixed in the same module?

Reply:
hehe, `zlistx_t` doesn't have some friendly functions like `zlist_pop()`.  Let me make it work with the `zlistx_t` interface.

Reply:
all of 30 seconds into trying to adjust to `zlistx_t`, I remember why I kept it as `zlist_t`.  `zlistx_t` uses handles for detach/delete.  But since this list is for keeping track of futures and not my own personal data structure, I'd have to do `flux_future_aux_set()` to save off the handle in the future.  Not sure if it's a net win.  I'll code it up and see.

Reply:
`zlist_remove()` searches the list for the item, so the equivalent in zlistx would just be
```c
void *handle = zlistx_find (futures, f);
if (handle)
    zlistx_delete (futures, handle); // or zlistx_detach()
```

Reply:
ahh good idea.